extensions/arc/src/ui/dialogs/connectControllerDialog.ts

```diff
@@ -5,37 +5,73 @@
 
 import { ControllerInfo } from 'arc';
 import * as azdata from 'azdata';
+import * as azdataExt from 'azdata-ext';
 import { v4 as uuid } from 'uuid';
 import * as vscode from 'vscode';
 import { Deferred } from '../../common/promise';
 import * as loc from '../../localizedConstants';
 import { ControllerModel } from '../../models/controllerModel';
 import { InitializingComponent } from '../components/initializingComponent';
 import { AzureArcTreeDataProvider } from '../tree/azureArcTreeDataProvider';
+import { getErrorMessage } from '../../common/utils';
 
 export type ConnectToControllerDialogModel = { controllerModel: ControllerModel, password: string };
 export class ConnectToControllerDialog extends InitializingComponent {
 	private modelBuilder!: azdata.ModelBuilder;
+	protected dialog: azdata.window.Dialog;
 
-	private urlInputBox!: azdata.InputBoxComponent;
-	private nameInputBox!: azdata.InputBoxComponent;
-	private usernameInputBox!: azdata.InputBoxComponent;
-	private passwordInputBox!: azdata.InputBoxComponent;
-	private rememberPwCheckBox!: azdata.CheckBoxComponent;
+	protected urlInputBox!: azdata.InputBoxComponent;
+	protected nameInputBox!: azdata.InputBoxComponent;
+	protected usernameInputBox!: azdata.InputBoxComponent;
+	protected passwordInputBox!: azdata.InputBoxComponent;
+	protected rememberPwCheckBox!: azdata.CheckBoxComponent;
 
-	private _completionPromise = new Deferred<ConnectToControllerDialogModel | undefined>();
+	protected _completionPromise = new Deferred<ConnectToControllerDialogModel | undefined>();
+	protected _id!: string;
 
-	private _id!: string;
 
-	constructor(private _treeDataProvider: AzureArcTreeDataProvider) {
+	constructor(protected _treeDataProvider: AzureArcTreeDataProvider, title: string = loc.connectToController) {
 		super();
+		this.dialog = azdata.window.createModelViewDialog(title);
+	}
+
+	private getComponents(isPwReacquisition: boolean): (azdata.FormComponent<azdata.Component> & { layout?: azdata.FormItemLayout | undefined; })[] {
```

Comment:
For what it's worth I don't like how this is just increasing the complexity of the base class and adding a bunch of extra code for very little benefit. I personally would have just created a new dialog class that was very similar looking to this one - they share the basic stuff in common but there's enough custom logic and overrides here that it's just more confusing and harder to follow both sides now.

I guess it's ok to leave like this but please try to take stuff like that into account in the future.

Reply:
Agreed. I do not like how this turned out code-wise. I will clean it up in a future PR. #12294